server/models/dataset.py

```diff
@@ -294,17 +300,218 @@ def addImage(self, dataset, imageDataStream, imageDataSize, filename, signature,
+        uploadRoleArn = Setting().get(
+            PluginSettings.ZIP_UPLOAD_ROLE_ARN)
+        assumeRoleDurationSeconds = Setting().get(
+            PluginSettings.ZIP_UPLOAD_ASSUME_ROLE_DURATION_SECONDS)
+        if not all([accessKeyId, secretAccessKey, s3BucketName, uploadRoleArn,
+                    assumeRoleDurationSeconds]):
+            raise GirderException('ZIP upload not configured.')
+
+        # Create new batch
+        batch = Batch().createBatch(
+            dataset=dataset,
+            creator=user,
+            signature=signature)
+
+        # Add policy that restricts uploads to only the specific key
+        s3BucketArn = 'arn:aws:s3:::%s' % s3BucketName
+        s3ObjectKey = 'zip-uploads/%s' % batch['_id']
+        s3BucketPutObjectInKeyPolicy = {
+            "Version": "2012-10-17",
+            "Statement": [
+                {
+                    "Effect": "Allow",
+                    "Action": [
+                        "s3:PutObject",
+                    ],
+                    "Resource": "%s/%s" % (s3BucketArn, s3ObjectKey)
+                }
+            ]
+        }
+
+        # Get temporary security credentials with permission to upload into the
+        # object in the S3 bucket. The AWS Security Token Server (STS) provides
+        # the credentials when the upload user assumes the upload role.
+        sts = boto3.client(
+            'sts',
+            aws_access_key_id=accessKeyId,
+            aws_secret_access_key=secretAccessKey)
+        try:
+            resp = sts.assume_role(
+                RoleArn=uploadRoleArn,
+                RoleSessionName='ZipUploadSession-%d' % int(time.time()),
+                Policy=json.dumps(s3BucketPutObjectInKeyPolicy),
+                DurationSeconds=assumeRoleDurationSeconds
+            )
+        except botocore.exceptions.ClientError as e:
+            raise GirderException('Error acquiring temporary security credentials: %s' %
+                                  e.response['Error']['Message'])
+
+        # TODO: Could store assumed role user ARN on batch for use as principal in a bucket policy
+        # that effectively revokes the temporary security credentials, before they expire,
+        # on cancel/finalize:
+        # https://docs.aws.amazon.com/IAM/latest/UserGuide/id_credentials_temp_control-access_disable-perms.html
+        # assumedRoleUserArn = resp['AssumedRoleUser']['Arn']
+
+        credentials = resp['Credentials']
+
+        # Store upload information on batch
+        batch.update({
+            's3BucketName': s3BucketName,
+            's3ObjectKey': s3ObjectKey
+        })
+        batch = Batch().save(batch)
+
+        return {
+            'accessKeyId': credentials['AccessKeyId'],
+            'secretAccessKey': credentials['SecretAccessKey'],
+            'sessionToken': credentials['SessionToken'],
+            'bucketName': s3BucketName,
+            'objectKey': s3ObjectKey,
+            'zipId': batch['_id']
```

Comment:
Maybe we can just call this `batchId` to prevent having different internal / external names.